optking/optimize.py

```diff
@@ -27,364 +27,313 @@ def optimize(o_molsys, computer):
+            )
+            raise OptError(
+                "Maximum number of steps exceeded: {}.".format(self.params.geom_maxiter), "OptError",
+            )
+
+    def get_hessian_protocol(self):
+        """Determine action to take for how to compute a hessian. Handles alternate IRC behavior
+
+        Returns
+        -------
+        str: one of ('compute', 'update', 'guess', 'unneeded')
+
+        """
+
+        if "hessian" not in self.opt_method.requires():
+            return "unneeded"
+
+        if self.erase_hessian is True:
+            self.erase_hessian = False
+            return "guess"
+
+        if self.step_number <= 1:
+            if self.params.opt_type != "IRC":
+                if self.params.full_hess_every > -1:  # compute hessian at least once.
+                    protocol = "compute"
+                else:
+                    protocol = "guess"
+            else:  # IRC
+                protocol = "compute"
+        else:
+            if self.params.full_hess_every < 1:
+                protocol = "update"
+            elif self.step_number % self.params.full_hess_every == 0:
+                protocol = "update"
+            else:
+                protocol = "update"
+
+        self.logger.info(protocol)
+        return protocol
+
+    def clear(self):
+        self.history.steps = []
+        self.molsys.intcos = []
+        self.step_number = 0
+        self.history.steps_since_last_hessian = 0
+        self.history.consecutive_backsteps = 0
+
+    def alg_error_handler(self, error):
+
+        self.logger.error("\n\tCaught AlgError exception\n")
+        eraseIntcos = False
+
+        if error.linearBends:
+            # New linear bends detected; Add them, and continue at current level.
+            # from . import bend # import not currently being used according to IDE
+            for l in error.linearBends:
+                if l.bend_type == "LINEAR":  # no need to repeat this code for "COMPLEMENT"
+                    iF = addIntcos.check_fragment(l.atoms, self.molsys)
+                    F = self.molsys.fragments[iF]
+                    intcosMisc.remove_old_now_linear_bend(l.atoms, F.intcos)
+                    F.add_intcos_from_connectivity()
+            eraseHistory = True
+        elif op.Params.dynamic_level == op.Params.dynamic_level_max:
+            self.logger.critical("\n\t Current algorithm/dynamic_level is %d.\n" % op.Params.dynamic_level)
+            self.logger.critical("\n\t Alternative approaches are not available or turned on.\n")
+            raise OptError("Maximum dynamic_level reached.")
+        else:
+            op.Params.dynamic_level += 1
+            self.logger.warning("\n\t Increasing dynamic_level algorithm to %d.\n" % op.Params.dynamic_level)
+            self.logger.warning("\n\t Erasing old history, hessian, intcos.\n")
+            eraseIntcos = True
+            eraseHistory = True
+            op.Params.updateDynamicLevelParameters(op.Params.dynamic_level)
```

Comment:
Are dynamic level parameters getting used now, or do we need another loop over them?  We need this to handle random users trying hard cases; cases that other optimizers can't do.